client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +159,127 @@ public void start(Listener listener) {
-                        .addAll(controllerServerList.getNodeURIList()
-                                .stream()
-                                .map(node ->
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                final ResolvedServerInfoGroup serverInfoGroup;
+                if (this.enableDiscovery) {
+                    // Make an RPC call to the bootstrapped controller servers to fetch all active controllers.
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server hostnames to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
+                        final InetSocketAddress socketAddress = new InetSocketAddress(address.getHostString(),
+                                address.getPort());
+                        if (!socketAddress.isUnresolved()) {
+                            resolvedAddresses.add(socketAddress);
+                        }
+                    });
+                    serverInfoGroup = ResolvedServerInfoGroup.builder().addAll(resolvedAddresses.stream()
+                            .map(ResolvedServerInfo::new)
+                            .collect(Collectors.toList()))
+                            .build();
+                }
 
                 // Update gRPC load balancer with the new set of server addresses.
                 log.info("Updating client with controllers: {}", serverInfoGroup);
                 this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
-            } catch (StatusRuntimeException e) {
-                log.warn("Failed to fetch controller addresses - {}", e);
-                this.resolverUpdater.onError(e.getStatus());
+
+                // We have found at least one controller endpoint. Repeat discovery after the regular schedule.
+                this.scheduledFuture = this.scheduledExecutor.schedule(
```

Comment:
Yes thats another way to do it and makes sense when we use external executor (https://github.com/pravega/pravega/issues/1546). With the dedicated thread model current implementation is much simpler.